src/main/java/in/projecteka/consentmanager/user/PatientsController.java

```diff
@@ -59,22 +72,104 @@
                 .map(securityContext -> (Caller) securityContext.getAuthentication().getPrincipal())
                 .flatMap(caller -> transactionPinService.validatePinFor(caller.getUsername(),
-                request.getPin(),
-                request.getRequestId(),
-                request.getScope()));
+                        request.getPin(),
+                        request.getRequestId(),
+                        request.getScope()));
     }
 
     @PostMapping("/profile")
     public Mono<Session> create(@RequestBody SignUpRequest request,
                                 @RequestHeader(name = "Authorization") String token) {
         var signUpRequests = SignUpRequestValidator.validate(request, userService.getUserIdSuffix());
         return signUpRequests.isValid()
-               ? Mono.justOrEmpty(signupService.sessionFrom(token))
-                       .flatMap(sessionId -> userService.create(signUpRequests.get(), sessionId)
-                               .zipWith(Mono.just(sessionId))
-                               .flatMap(tuple -> signupService.removeOf(tuple.getT2()).thenReturn(tuple.getT1())))
-               : Mono.error(new ClientError(BAD_REQUEST,
-                       new ErrorRepresentation(new Error(INVALID_REQUESTER,
-                               signUpRequests.getError().reduce((left, right) -> format("%s, %s", left, right))))));
+                ? Mono.justOrEmpty(signupService.sessionFrom(token))
+                .flatMap(sessionId -> userService.create(signUpRequests.get(), sessionId)
+                        .zipWith(Mono.just(sessionId))
+                        .flatMap(tuple -> signupService.removeOf(tuple.getT2()).thenReturn(tuple.getT1())))
+                : Mono.error(new ClientError(BAD_REQUEST,
+                new ErrorRepresentation(new Error(INVALID_REQUESTER,
+                        signUpRequests.getError().reduce((left, right) -> format("%s, %s", left, right))))));
+    }
+
+    @PostMapping("/generateotp")
+    @ResponseStatus(CREATED)
+    public Mono<GenerateOtpResponse> generateOtp(@RequestBody GenerateOtpRequest request) {
+        return profileService.profileFor(request.getUsername())
+                .switchIfEmpty(Mono.error(ClientError.userNotFound()))
+                .flatMap(this::getVerfiedIdentifier)
+                .switchIfEmpty(Mono.error(ClientError.unAuthorized()))
+                .flatMap(verifiedIdentifier -> getGenerateOtpResponse(verifiedIdentifier, request.getUsername()))
+                .switchIfEmpty(Mono.error(ClientError.failedToGenerateOtp()));
+    }
+
+    @PostMapping("/verifyotp")
+    public Mono<Token> verifyOtp(@RequestBody OtpVerification request) {
+        return userService.verifyOtp(request);
+    }
+
+    private Mono<GenerateOtpResponse> getGenerateOtpResponse(UserSignUpEnquiry userSignUpEnquiry, String userName) {
+        SignUpSession signUpSession = userService.sendOtpForPasswordChange(userSignUpEnquiry, userName).block();
```

Comment:
Is it fine to block the call here? I am not sure.

Reply:
Yes. We are not using the signUpSession post this. Only after blocking can we get the sessionId

Reply:
This code can simply be following without blocking,
```
userService.sendOtpForPasswordChange(userSignUpEnquiry, userName)
                    .flatMap(signUpSession -> {
                                    OtpMediumType otpMediumType = getOtpMedium(userSignUpEnquiry);
                                    String maskedIdentifier = maskedIdentifier(userSignUpEnquiry);

                                    if (signUpSession == null || otpMediumType == null || maskedIdentifier == null) {
                                          return Mono.empty();
                                    }
                                    return Mono.just(GenerateOtpResponse.builder()
                                                   .sessionId(signUpSession.getSessionId())
                                                   .otpMediumValue(maskedIdentifier)
                                                   .otpMedium(otpMediumType.toString())
                                                   .expiryInMinutes(userService.getExpiryInMinutes())
                                                   .build());
});
```

Reply:
Note: Formatting may not be correct.

Reply:
Done